### PR TITLE
Fix error/warnings for config file

### PR DESCRIPTION
**Pull request by H-ANSEN**
The following pull request seeks to resolve the issue listed below!

## Expected Behavior
The configuration file for dune makes use of the same version numbering scheme as dune-project files *`(lang dune <ver>)`*. When attempting to use dune with a configuration file that has a mismatch between a stanza and it's supported dune version an error indicating this mismatch in the configuration file should be displayed:
```cmd
File "dune/config", line 2, characters 0-15:
2 | (cache enabled)
Error: 'cache' is only available since version 2.0 of the dune language.
Please update your dune config file to have (lang dune 2.0).
```
## Actual Behavior
The actual recommends changing the version in the `dune-project` file as opposed to the users config file:
```cmd
File "dune/config" line 2, characters 0-15:
2 | (cache enabled)
Error: 'cache' is only available since version 2.0 of the dune language.
Please update your dune-project file to have (lang dune 2.0).
```
## Reproduction
```cmd
printf "(lang dune 1.0)\n(cache enabled)" > dune-config
dune init proj test --config-file=dune-config
```

### CODE REVIEW

File: doc/changes/10923.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+- Fix error/warning messages displayed when the specified dune configuration
```
Comment (Leonidas-from-XIV) [inline doc/changes/10923.md:1]: I would say what it fixes is the file that the error refers to, maybe worth being explicit about it in the changelog.

File: src/dune_sexp/syntax.ml [hunk 2]
```diff
@@ -250,15 +250,28 @@ module Key = struct
...
       what
       (Version.to_string ver)
       t.desc
-      lang_or_using
+      (lang_or_using t.name)
+      t.name
+      (Version.to_string ver)
+  ;;
+
+  let since_config t ver ~what =
```
Comment (Leonidas-from-XIV) [inline src/dune_sexp/syntax.ml:267]: Is there a point in having this be two separate functions? To me it seems like what it does is

```ocaml
sprintf "%s is only available since version %s of %s. Please update your %s file to have (%s %s %s)."

i.e. the file to be mentioned can also just be an argument.
```

File: src/dune_sexp/syntax.mli [hunk 3]
```diff
@@ -71,6 +75,8 @@ val create
 (** Return the name of the syntax. *)
 val name : t -> string
 
+val desc : unit -> (Loc.t * string, _) Decoder.parser
```
Comment (Leonidas-from-XIV) [inline src/dune_sexp/syntax.mli:78]: Can you add the type of `_` here as well? `_` is useful for writing code but looking at an `mli` file `_` is really unhelpful. Also a short docstring would be appreciated.

File: src/dune_sexp/syntax.mli [hunk 4]
```diff
@@ -98,6 +104,16 @@ val renamed_in : t -> Version.t -> to_:string -> (unit, _) Decoder.parser
...
 
+(** Like [since] but accepts a function [fmt] allowing custom formatting of the
+    entire error/warning message. See [Error_msg] for format functions. *)
+val since_fmt
+  :  ?fatal:bool
+  -> fmt:(t -> Version.t -> string)
+  -> t
+  -> Version.t
+  -> Loc.t
+  -> (unit, _) Decoder.parser
```
Comment (Leonidas-from-XIV) [inline src/dune_sexp/syntax.mli:115]: Likewise here for `_`.

File: test/blackbox-tests/test-cases/config/config-version.t [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+Attempt to initialize a new project while an invaild stanza due to versioning
+exists in the config.
+
+  $ dune init proj test --config-file=dune-config
+  File "$TESTCASE_ROOT/dune-config", line 2, characters 0-15:
+  2 | (cache enabled)
+      ^^^^^^^^^^^^^^^
+  Error: 'cache' is only available since version 2.0 of the dune language.
+  Please update your dune config file to have (lang dune 2.0).
+  [1]
```
Comment (Leonidas-from-XIV) [inline test/blackbox-tests/test-cases/config/config-version.t:23]: Can you add a test that adds `(lang dune 2.0)` and shows this works? There is probably a test that makes sure of it, but it is also useful to show that it works.

File: test/blackbox-tests/test-cases/config/config-version.t [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+Swap out the invaild stanza for another and try again.
+
+  $ sed -i -e '$s/.*/(cache-check-probability 0.5)/' dune-config
```
Comment (Leonidas-from-XIV) [inline test/blackbox-tests/test-cases/config/config-version.t:27]: This can be done with `echo "(cache-check-probability 0.5)" >> dune-config` which is a bit simpler to read and achieves a similar result.

File: test/blackbox-tests/test-cases/config/config-version.t [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  $ dune init proj test --config-file=dune-config
+  File "$TESTCASE_ROOT/dune-config", line 2, characters 0-29:
+  2 | (cache-check-probability 0.5)
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  Error: 'cache-check-probability' is only available since version 2.7 of the
+  dune language. Please update your dune config file to have (lang dune 2.7).
+  [1]
+
+Swap out the invaild stanza for another and try again.
```
Comment (Leonidas-from-XIV) [inline test/blackbox-tests/test-cases/config/config-version.t:36]: Correct me if I'm misunderstanding something but past this point I don't see a purpose of the other tests. We already established that it displays the right file when it errors and it works when deeding the advice and adding a new error shows the error message again with the right error location.

I'd remove the rest of the tests as they don't exercise any more interesting edge cases (and running extra tests takes time that could be better spent).

**Review by Leonidas-from-XIV — CHANGES_REQUESTED** [review]
Thanks for the contribution, I appreciate the extensive tests. The code is fine, see my comments on what would be worth improving to get it merged.

File: src/dune_sexp/syntax.ml (hunk 2 above)
Comment (H-ANSEN) [reply src/dune_sexp/syntax.ml:267]: Yea! Everything here can be super simplified if we take the file name as an argument. When deciding how to fix this my initial thought was to alter the `since` function as follows:
```ocaml
let since ?(file = "dune-project") ?(fatal = true) ?what t ver = ...
```
Due to the fact that `since` is used all over the place I felt that it might be worth proposing and implementation which preserved the original signature and also provided a way to control the entire error message. If you think that altering the signature as shown above would be okay I'd be happy to move towards this type of representation, this would also have the added benefit simplifying/resolving some of the other changes you mention below!

File: test/blackbox-tests/test-cases/config/config-version.t (hunk 7 above)
Comment (H-ANSEN) [reply test/blackbox-tests/test-cases/config/config-version.t:36]: Your correct here, my intention was to simply cover multiple stanzas with differing supported versions of dune but 2 tests does cover this case.

File: test/blackbox-tests/test-cases/config/config-version.t (hunk 6 above)
Comment (H-ANSEN) [reply test/blackbox-tests/test-cases/config/config-version.t:27]: Here I am using sed to **swap** out the first invalid stanza *'(cache enabled)'*. `>>` has the effect of appending rather than swapping so the resulting error would be caused on the same *'(cache enabled)'* stanza from the first test instead of the new *cache-check-probability* stanza!

File: src/dune_sexp/syntax.ml (hunk 2 above)
Comment (H-ANSEN) [reply src/dune_sexp/syntax.ml:267]: Your correct! It's functionally the same thing just with a different string describing a different file. I can condense this into a single function.

File: test/blackbox-tests/test-cases/config/config-version.t (hunk 6 above)
Comment (H-ANSEN) [reply test/blackbox-tests/test-cases/config/config-version.t]: Ahh sorry, I did not understand what you initially meant, thank you for clarifying! I've cleaned up the test cases let me know what you think.

**Review by Leonidas-from-XIV — CHANGES_REQUESTED** [review]
Thanks for the revisions, the tests look pretty good now. I have one clarification and one small thing to make the tests safer, after that I think it's good to go!

File: test/blackbox-tests/test-cases/config/config-version.t (hunk 6 above)
Comment (Leonidas-from-XIV) [reply test/blackbox-tests/test-cases/config/config-version.t]: They have a good flow demonstrating the issue with the error and the fix, I like it.

**Review by Leonidas-from-XIV — APPROVED** [review]
Looks good, thanks!
